### PR TITLE
Refactor to support facts and config reports

### PR DESCRIPTION
**Pull request by ekohl**
This is needed to test the foreman callback plugin. Currently I still need to look into the exact API responses and handling so it matches the Foreman.

### CODE REVIEW

File: foreman_stub.py [hunk 1]
```diff
@@ -9,61 +9,146 @@
...
 
-from flask import Flask, jsonify, request
+from flask import Flask, abort, jsonify, request
 
 
 DEBUG = os.getenv('DEBUG', 'false').lower() not in ('false', '0', '', 'off')
 
 if DEBUG:
-    import pip
-    pip.main(['install', 'q', 'epdb'])
```
Comment (webknjaz) [inline foreman_stub.py:19]: I don't think it's needed

File: foreman_stub.py [hunk 2]
```diff
@@ -9,61 +9,146 @@
...
     for jfile in jfiles:
-        with open(jfile, 'r') as f:
-            jdata = json.load(f)
+        with open(jfile, 'r') as fixture:
+            jdata = json.load(fixture)
         key = urlparse(jdata['url']).path[7:]
         subkey = jdata['page']
         PAGECACHE[key][subkey] = jdata['data']
 
+        if 'id' in jdata['data'] and key == f"/hosts/{jdata['data']['id']}":
```
Comment (webknjaz) [inline foreman_stub.py:34]: Pythonic way would be to follow EAFP with try/except KeyError

File: foreman_stub.py [hunk 3]
```diff
@@ -9,61 +9,146 @@
...
+
 
 def get_page_num():
-    """Return page number if present."""
-    pagenum = request.args.get('page')
-    if pagenum:
-        pagenum = int(pagenum)
-    return pagenum
+    """Return page number."""
+    return int(request.args.get('page', 1))
```
Comment (webknjaz) [inline foreman_stub.py]: Wouldn't it crash on an empty string?

File: foreman_stub.py (hunk 2 above)
Comment (ekohl) [reply foreman_stub.py:34]: Right now you can got to `/hosts`, find a host and `/hosts/ID` will return a 404 which meant it took me some time to find a valid host when testing in the callback plugin.

This is an ugly workaround to have both `/hosts/1` and `/hosts/host.example.com` work but it's really a partial solution. Looking at the data set it might be easier to have a plain list of hosts and pretend it's a database.

What do you think about changing the `PAGECACHE` to `HOSTS = []` and appending all results? I'm already imitating the pagination elsewhere anyway.

File: foreman_stub.py (hunk 3 above)
Comment (ekohl) [reply foreman_stub.py]: You're right, but you just shouldn't do that ;) I'll modify this.

File: foreman_stub.py [hunk 4]
```diff
@@ -9,61 +9,146 @@
...
+    host = find_host(hostid)
+
+    if request.method == 'POST':
+        host['facts'] = request.get_json()
+
+    page = get_page_num()
+    per_page = get_per_page()
+
+    start = (page - 1) * per_page
+    end = (page) * per_page
```
Comment (webknjaz) [inline foreman_stub.py]:
```suggestion
    end = page * per_page
```

File: foreman_stub.py [hunk 5]
```diff
@@ -9,61 +9,146 @@
...
+def get_host(hostid):
+    """Render fixture contents from cache."""
+    return jsonify(find_host(hostid))
+
+
+@app.route('/api/v2/hosts/<hostid>/facts', methods=['GET'])
+def get_facts(hostid):
+    host = find_host(hostid)
+
+    if request.method == 'POST':
```
Comment (webknjaz) [inline foreman_stub.py]: This route is only registered to receive GET quires. Line 89.

File: foreman_stub.py [hunk 6]
```diff
@@ -9,61 +9,146 @@
...
     except KeyError:
-        if DEBUG:
-            import q; q/cache_key; q/pagenum; q/hostid
-            import epdb; epdb.st()
-    else:
-        return jsonify(resp)
+        abort(400)
+
+    REPORTS.append(report)
+    resp = {}
```
Comment (webknjaz) [inline foreman_stub.py]: Extra var is definitely not needed here

File: foreman_stub.py [hunk 7]
```diff
@@ -72,4 +157,5 @@ def ping_heartbeat():
     return jsonify({"status": "ok", "response": "pong"})
 
 
-__name__ == '__main__' and app.run(host='0.0.0.0', port=int(os.getenv('PORT', 8080)), threaded=True, debug=DEBUG)
+if __name__ == '__main__':
```
Comment (webknjaz) [inline foreman_stub.py:185]: This is completely unrelated. And unneeded. Makes it hard to review.

File: foreman_stub.py [hunk 8]
```diff
@@ -9,61 +9,146 @@
...
+def get_config_reports():
+    page = get_page_num()
+    per_page = get_per_page()
+
+    start = (page - 1) * per_page
+    end = (page) * per_page
+
+    resp = {
+        'total': len(REPORTS),
+        'subotal': len(REPORTS),
```
Comment (webknjaz) [inline foreman_stub.py]: You count length twice here.

File: foreman_stub.py [hunk 9]
```diff
@@ -9,61 +9,146 @@
...
+        'total': len(host['facts']),
+        'subotal': 1,  # this appears to always be 1 in the API
+        'page': page,
+        'per_page': per_page,
+        'sort': {
+            'by': None,
+            'order': None,
+        },
+        'results': {
+            host['name']: dict(sorted(host['facts'].items())[start:end]),
```
Comment (webknjaz) [inline foreman_stub.py:107]: How would tuples be sorted? Is the first item used?

File: foreman_stub.py (hunk 2 above)
Comment (webknjaz) [reply foreman_stub.py:34]: I think that refactoring and actual changes should be separated. I cannot review this when there's more unrelated changes than justified ones.

Comment (webknjaz) [reply foreman_stub.py:34]: As for the check, I assume you meant something like
```suggestion
        if key == f"/hosts/{jdata['data'].get('id') or ''}":
```

File: foreman_stub.py (hunk 3 above)
Comment (webknjaz) [reply foreman_stub.py]: I think that I want only related changes in this PR. It'll shrink it to the reviewable size. Such things are just a distraction.

File: foreman_stub.py [hunk 10]
```diff
@@ -9,61 +9,146 @@
...
+    resp = PAGECACHE['/hosts'][pagenum]
+    return jsonify(resp)
 
-    cache_key = '/hosts'
-    if hostid is not None:
-        cache_key = f'{cache_key}/{hostid}'
 
-    resp = PAGECACHE[cache_key]
+@app.route('/api/v2/hosts/facts', methods=['POST'])
+def handle_facts():
```
Comment (webknjaz) [inline foreman_stub.py:65]: Please follow the tradition of adding docstrings to functions.

File: foreman_stub.py (hunk 8 above)
Comment (ekohl) [reply foreman_stub.py]: There is a mistake in this line, but it's a typo in subtotal. The reason this makes sense here is that subtotal is only smaller when you have a search string. This isn't implemented in the dummy so they always are the same.

File: foreman_stub.py (hunk 9 above)
Comment (ekohl) [reply foreman_stub.py:107]: Yes.

File: foreman_stub.py (hunk 10 above)
Comment (ekohl) [reply foreman_stub.py:65]: It's still a draft :)

File: foreman_stub.py (hunk 6 above)
Comment (ekohl) [reply foreman_stub.py]: This is why it's a draft: I still need to look into the exact response.

File: foreman_stub.py (hunk 3 above)
Comment (ekohl) [reply foreman_stub.py]: Actually, the default value of 1 change is needed for the host detail page to work correctly.

File: foreman_stub.py (hunk 8 above)
Comment (webknjaz) [reply foreman_stub.py]: I mean that you do two invocations of `len()` which is suboptimal :)

File: foreman_stub.py (hunk 3 above)
Comment (webknjaz) [reply foreman_stub.py]: well, then maybe
```suggestion
    return int(request.args.get('page', 1) or 1)
```

File: foreman_stub.py (hunk 8 above)
Comment (ekohl) [reply foreman_stub.py]: Fair point :)

File: foreman_stub.py (hunk 3 above)
Comment (ekohl) [reply foreman_stub.py]: I was thinking about that, but `return int(request.args.get('page') or 1)` is probably sufficient.

File: foreman_stub.py (hunk 2 above)
Comment (ekohl) [reply foreman_stub.py:34]: I think that's incorrect because if there's no ID, there is probably also no hostname. Otherwise I think you'll be mapping lists to individual data pages.

File: foreman_stub.py (hunk 7 above)
Comment (ekohl) [reply foreman_stub.py:185]: It's a separate commit. It was useful for me to have a pylint/pyflakes clean file so my editor makes it easier to spot bugs I'm introducing.

**Comment by ekohl** [comment]
Should be good to review now. My verification shows this is now good enough. Long term I'd still think about a `HOSTS = []` structure so you can simulate a simple database.

**Review by webknjaz — CHANGES_REQUESTED** [review]
epdb removal must be reverted.
